File: 20200215Python-China/0096. Unique Binary Search Trees.py
'''
Given n, how many structurally unique BST's (binary search trees) that store values 1 ... n?

Example:

Input: 3
Output: 5
Explanation:
Given n = 3, there are a total of 5 unique BST's:

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
'''
class Solution:
    def numTrees(self, n: int) -> int:
        if n==0:
            return 0
        if n==1:
            return 1
        dp=[1,1]
        for k in range(2,n+1):
            cur=0
            for i in range(k):
                cur+=dp[i]*dp[k-1-i]
            dp.append(cur)
        return dp[-1]

File: archives/C#/0096. Unique Binary Search Trees.cs
public class Solution {
    public int NumTrees(int n) {
        /*
        1   1
        2   f(1)+f(1)=2
        3   f(0)f(2)+f(1)f(1)+f(2)f(0)=5
        4   f(0)(3)+f(1)f(2)+f(2)f(1)+f(3)f(0)=...
        5   
        f(n)=f(0)f(n-1)+f(1)f(n-1-1)+...+f(i)f(n-1-i)+...+f(n-1)f(0)
        */
        
        List<int> dp=new List<int>();
        dp.Add(1);
        dp.Add(1);
        for(int i=2;i<n+1;i++){
            int tmp=0;
            for(int j=0;j<i;j++){
                tmp+=dp[j]*dp[i-1-j];
            }
            dp.Add(tmp);
        }
        return dp[n];
    }
}
